config: add enable-email-login setting

mhilton opened the pull request:
Add a new configuration setting to enable the "login with email"
functionality, making it disabled by default.

jujugui commented:
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/candid/61/

jujugui commented:
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/candid/63/

jujugui commented:
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/candid/64/

jujugui commented:
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/candid/65/

jujugui commented:
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/candid/66/

mhilton commented on store/keyvalue.go:
```diff
@@ -35,14 +36,35 @@ type KeyValueStore interface {
+// SetKeyOnce is like KeyValueStore.Set except that if the key already
+// has a value associated with it it returns an error with the cause of
+// ErrDuplicateKey.
+func SetKeyOnce(ctx context.Context, kv KeyValueStore, key string, value []byte, expire time.Time) error {
```
I wonder if this would be better expressed as a getVal func. How about

```
// SetOnce creates a function that can be used with Update to set it to the given value if it is not set.
func SetOnce(val []byte) func([]byte) ([]byte, error) {
	return func(oldVal) ([]byte, error) {
		if oldVal == nil {
			return val, nil
		}
		return oldVal, nil
	}
}
```

mhilton commented on store/memstore/keyvalue.go:
```diff
@@ -63,18 +64,24 @@ func (s keyValueStore) Get(_ context.Context, key string) ([]byte, error) {
-	_, ok := s.store.data[s.idp][key]
-	if ok {
-		return store.DuplicateKeyError(key)
+	data := s.store.data[s.idp]
```
Is this reasonable behavior to use the current value if it should have expired? The definition of expiry is quite loose but even so it might be confusing.

rogpeppe replied:
Yeah, I think it's fine. The current code ignores expiry times entirely and other backends only expire after a while.

mhilton commented on store/sqlstore/backend.go:
```diff
@@ -110,6 +110,7 @@ const (
 	tmplPushIdentitySet
 	tmplPullIdentitySet
 	tmplGetProviderData
+	tmplGetProviderDataForUpdate
```
I'm slightly worried that this is a bit too postgresql specific an operation name, probably not a problem as we don't currently support anything else.

rogpeppe replied:
It may be, but we can always adjust things later if that turns out to be the case.

mhilton commented on store/sqlstore/keyvalue.go:
```diff
@@ -48,14 +48,28 @@ type providerDataParams struct {
+	return v, nil
+}
+
+// Get implements store.KeyValueStore.Get by selecting the blob with the
```
Please update this doc comment

mhilton commented on store/sqlstore/keyvalue.go:
```diff
@@ -48,14 +48,28 @@ type providerDataParams struct {
+
+// Get implements store.KeyValueStore.Get by selecting the blob with the
+// given key from the provider_data table.
+func (s *keyValueStore) get(_ context.Context, q queryer, key string, forUpdate bool) ([]byte, error) {
```
isit really worth including the context parameter that is ignored in an internal method?

mhilton commented on store/sqlstore/keyvalue.go:
```diff
@@ -79,10 +97,32 @@ func (s *keyValueStore) Set(_ context.Context, key string, value []byte, expire
 	return errgo.Mask(err)
 }
 
+func (s *keyValueStore) Update(ctx context.Context, key string, expire time.Time, getVal func(old []byte) ([]byte, error)) error {
```
Please add a doc comment

rogpeppe commented:
:shipit:

jujugui commented:
Status: merge request accepted. Url: http://ci-gce.theblues.io:8080/job/candid-merge

jujugui commented:
Build failed: Attempt to land pull request failed
        build url: http://ci-gce.theblues.io:8080/job/candid-merge/31

jujugui commented:
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/candid/67/

rogpeppe commented:
:shipit:

jujugui commented:
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/candid/70/

jujugui commented:
Status: merge request accepted. Url: http://ci-gce.theblues.io:8080/job/candid-merge

jujugui commented:
Build failed: Attempt to land pull request failed
        build url: http://ci-gce.theblues.io:8080/job/candid-merge/34

rogpeppe commented:
:shipit:

jujugui commented:
Status: merge request accepted. Url: http://ci-gce.theblues.io:8080/job/candid-merge

jujugui commented:
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/candid/71/

jujugui commented:
Build failed: Attempt to land pull request failed
        build url: http://ci-gce.theblues.io:8080/job/candid-merge/35

rogpeppe commented:
:shipit:

jujugui commented:
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/candid/72/

jujugui commented:
Status: merge request accepted. Url: http://ci-gce.theblues.io:8080/job/candid-merge

mhilton committed:
config: add enable-email-login setting

Add a new configuration setting to enable the "login with email"
functionality, making it disabled by default.
